src/lib/homestore_backend/heap_chunk_selector.cpp

```diff
@@ -59,39 +55,21 @@ csharedChunk HeapChunkSelector::select_chunk(homestore::blk_count_t count, const
         return nullptr;
     }
 
-    std::shared_lock lock_guard(m_chunk_selector_mtx);
-    // FIXME @Hooper: Temporary bypass using pdev_id_hint to represent pg_id_hint, "identical layout" will change it
     pg_id_t pg_id = 0;
-    auto& pg_id_hint = hint.pdev_id_hint;
-    if (!pg_id_hint.has_value()) {
+    homestore::chunk_num_t chunk_id = 0;
+    if (!hint.pdev_id_hint.has_value()) {
         LOGWARNMOD(homeobject, "should not allocated a chunk without exiting pg_id in hint!");
         return nullptr;
     } else {
-        pg_id = pg_id_hint.value();
-    }
-
-    auto it = m_per_pg_heap.find(pg_id);
-    if (it == m_per_pg_heap.end()) {
-        LOGWARNMOD(homeobject, "No pg found for pg_id {}", pg_id);
-        return nullptr;
-    }
-
-    auto vchunk = VChunk(nullptr);
-    auto& heap = it->second->m_heap;
-    if (auto lock_guard = std::lock_guard< std::mutex >(it->second->mtx); !heap.empty()) {
-        vchunk = heap.top();
-        heap.pop();
+        // use pdev_id_hint (of type uint32_t) to store the values of pg_id and chunk_id.
+        // Both chunk_num_t and pg_id_t are of type uint16_t.
+        static_assert(std::is_same< pg_id_t, uint16_t >::value, "pg_id_t is not uint16_t");
+        static_assert(std::is_same< homestore::chunk_num_t, uint16_t >::value, "chunk_num_t is not uint16_t");
+        uint32_t pdev_id_hint = hint.pdev_id_hint.value();
+        pg_id = (uint16_t)(pdev_id_hint >> 16);
+        chunk_id = (uint16_t)(pdev_id_hint & 0xFFFF);
+        return select_specific_chunk(pg_id, chunk_id);
     }
-
-    if (vchunk.get_internal_chunk()) {
-        auto& avalableBlkCounter = it->second->available_blk_count;
-        avalableBlkCounter.fetch_sub(vchunk.available_blks());
-        remove_chunk_from_defrag_heap(vchunk.get_chunk_id());
-    } else {
-        LOGWARNMOD(homeobject, "no available chunks left for pg {}", pg_id);
-    }
-
-    return vchunk.get_internal_chunk();
 }
 
 csharedChunk HeapChunkSelector::select_specific_chunk(const pg_id_t pg_id, const chunk_num_t chunkID) {
```

Comment:
suggest use p_chunk_num_t and v_chunk_num_t to distinguish the expecting ID

Reply:
also suggesting if come with pg_id, always use vchunk_id (that belongs to PG).  using pgid + p_chunk_id is a bit wired .

Reply:
done